### ISSUE TITLE
Shoot Cluster ceases to work if kube-apiserver is blocked from metadata service access (AWS and GCP)

### ISSUE THREAD

**Issue by marwinski** [gardener/gardener#325]
When creating a PVC on a shoot cluster the kube-apiserver in the seed cluster tries to access the local metadata service. These are the requests: 

```
08:55:55.900520 IP kube-apiserver-5858674b87-7n9gv.42790 > metadata.google.internal.http: Flags [P.], seq 152:309, ack 294, win 30016, length 157: HTTP: GET /computeMetadata/v1/project/project-id HTTP/1.1
        0x0000:  4500 00c5 b288 4000 4006 9048 0af3 9872  E.....@.@..H...r
        0x0010:  a9fe a9fe a726 0050 6912 8c44 3c50 3e52  .....&.Pi..D<P>R
        0x0020:  5018 7540 f819 0000 4745 5420 2f63 6f6d  P.u@....GET./com
        0x0030:  7075 7465 4d65 7461 6461 7461 2f76 312f  puteMetadata/v1/
        0x0040:  7072 6f6a 6563 742f 7072 6f6a 6563 742d  project/project-
        0x0050:  6964 2048 5454 502f 312e 310d 0a48 6f73  id.HTTP/1.1..Hos
        0x0060:  743a 2031 3639 2e32 3534 2e31 3639 2e32  t:.169.254.169.2
        0x0070:  3534 0d0a 5573 6572 2d41 6765 6e74 3a20  54..User-Agent:.
        0x0080:  6763 6c6f 7564 2d67 6f6c 616e 672f 302e  gcloud-golang/0.
        0x0090:  310d 0a4d 6574 6164 6174 612d 466c 6176  1..Metadata-Flav
        0x00a0:  6f72 3a20 476f 6f67 6c65 0d0a 4163 6365  or:.Google..Acce
        0x00b0:  7074 2d45 6e63 6f64 696e 673a 2067 7a69  pt-Encoding:.gzi
        0x00c0:  700d 0a0d 0a                             p....
```

and 

```
HTTP: GET /computeMetadata/v1/instance/zone HTTP/1.1
```

Blocking access to metadata service causes the PV creation to fail that is created alongside the PVC. This is quite clearly a bug in Kubernetes as this information should be collected from the IaaS account for which credentials are provided.

**Comment by marwinski** [comment]
Update: we @zanetworker  @marwinski have just evaluated that the exact same problem does not show up on AWS.

**Comment by swapnilgm** [comment]
This issue has been already fixed in v0.9.0 via https://github.com/gardener/gardener/commit/50a3ec51b342e2a750045a7b8dc3192758186f14.

**Comment by zanetworker** [comment]
@swapnilgm  Thing is we don't need to disallow everywhere, it only happens on GCP based on our investigation. So we were going to look at the concealment service

**Comment by marwinski** [comment]
@swapnilgm: this hasn't been fixed and the Cloud Controller Service is not really here yet because it is beta and we might only be able to use it in some future versions (well and in addition we haven't tested it and it might not yet fully work as expected). 

Is it only a problem on GCP? I don't know. We haven't seen that specific problem on AWS but there might be others not to mention Azure / GCP / Alibaba. This said, we haven't really established whether the GCP metadata service is really dangerous, however when you look at what it contains and into the description of the metadata concealment service it most likely is dangerous. 

@swapnilgm: one more remark: have you actually tested the Kubernetes Cloud Controller Manager? It appears to depend one the same cloud provider code as the admission controller, so it might not solve the problem.

**Comment by marwinski** [comment]
According to the documentation at https://kubernetes.io/docs/reference/kubernetes-api/labels-annotations-taints/#failure-domainbetakubernetesiozone the labels are present on pv and node objects:

```
    failure-domain.beta.kubernetes.io/region: europe-west1
    failure-domain.beta.kubernetes.io/zone: europe-west1-b
```

In addition it also collects the project-id via 

```
curl http://169.254.169.254//computeMetadata/v1/project/project-id -H "Metadata-Flavor: Google"
sap-se-gcp-scp-k8s-dev
```

(see traced call above)

which makes it into the node spec of the nodes (but not the pv):

```
spec:
  podCIDR: 100.96.0.0/24
  providerID: gce://sap-se-gcp-scp-k8s-dev/europe-west1-b/shoot--core--afac-dev-worker-chpvg-z1-b9b7cb64f-k75jq
```

**Comment by marwinski** [comment]
It appears that the two call that are invoked on the metadata service (`computeMetadata/v1/project/project-id` and `/computeMetadata/v1/instance/zone` are not relevant for tagging the volumes. 

I have however found calls to the metadata service in the initialization of the gce cloud provider in `kubernetes/pkg/cloudprovider/providers/gce/gce.go` (in `generateCloudConfig`) invoked by the cloud provider initialization. This method invokes `getProjectAndZone` in `kubernetes/pkg/volume/gce_pd/gce_util.go` which goes directly to the metadata service.

This is kind of proven by the fact that the metadata service is invoked only upon the first creation of a PVC, not the second or third. Strange is however that the two methods named above are invoked twice on the first creation of a PVC.

So where does this leave us? I don't understand why the cloud provider needs the zone and project id but there might be a good reason for that so a pull request for the  cloud provider will likely break something and fail to be accepted.

On GCP there is the option to use the metadata concealer which @zanetworker  pointed out (https://cloud.google.com/kubernetes-engine/docs/how-to/metadata-concealment) and which might solve the problem.

I have seen lots of usage of metadata providers in the other cloud providers as well so the same problem that we do see on GCP might also exist for other cloud providers. 

This is my proposal on how to go forward:

- Run the kubernetes conformance tests on remaining cloud providers (with full monitoring of service the activity). When no traffic is detected run the tests again with the metadata service blocked. 

- If all cloud provider tests with blocked metadata service run successful use the GKE metadata concealer to protect GCP cluster otherwise we need to invest into our own concealer service.

**Comment by zanetworker** [comment]
@marwinski That was really helpful. It was actually @PadmaB  who suggested the use of the Metadata Concealment :)

**Comment by swapnilgm** [comment]
> have you actually tested the Kubernetes Cloud Controller Manager?

I haven't tested it out thoroughly. 

Just give you history, With initial discussion we started across restricting access from kube-apiserver to outside world, we considered blocking metadata service as well. I went ahead with blocking it, considering all the logic would be in kube-controller-manager and not in kube-apiserver. I did some basic test like deploying some core resources on shoot, restarting control-plane components for one of the cloud provider (surely not on GCP otherwise i had noticed that )  and everything well went.  
Yeah, I missed out on point that admission controller code might be a part of apiserver. @rfranzke l pointed the PVC issue on GCP and we decided to unblock the metadata service access from kube-apiserver.

Regarding you proposal about conformance test and GKE metadata concealer, I agree with you.
